utils/updateable_value.cc

```diff
@@ -93,30 +96,36 @@ updateable_value_base::updateable_value_base::operator=(std::nullptr_t) {
 
 void
 updateable_value_source_base::for_each_ref(std::function<void (updateable_value_base* ref)> func) {
-    for (auto ref : _refs) {
-        func(ref);
+    for (auto refs_per_shard : _refs) {
+        for (auto ref : refs_per_shard) {
+            func(ref);
+        }
```

Comment:
`for_each_ref` is used in `updateable_value_source::set` to propagate the new value to all referencing `updateable_value`s.

If your source is on shard 0 and your value is on shard 1, you can have a race between `updateable_value_source` propagating new T, and `updateable_value` reading its T. It's especially dangerous if T is not a simple type (e.g. unordered_map).

Reply:
Yes. That means .set() must return a future and use submit_to().

Reply:
True, I kind of assumed simple types, but there's no concept preventing complex types to be used here. Translating only `set()` to futures sounds like a nice compromise though, since all the constructors, destructors and operators remain the same. I'll give it a try in v2.